File: src/main/java/myratpackexamples/webpoll/PollRepository.java
package myratpackexamples.webpoll;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mongodb.async.client.MongoClient;
import com.mongodb.async.client.MongoClients;
import com.mongodb.async.client.MongoCollection;
import com.mongodb.async.client.MongoDatabase;
import io.vavr.control.Validation;
import myratpackexamples.webpoll.RatpackMongoClient.FindOneError;
import myratpackexamples.webpoll.RatpackMongoClient.InsertOneError;
import org.bson.Document;
import ratpack.exec.Promise;

import java.util.List;

import static ratpack.exec.Promise.error;

public class PollRepository {
    private ObjectMapper objectMapper = new ObjectMapper();

    public Promise<Validation<InsertOneError, Poll>> storePoll(PollRequest pollRequest) {
        try {
            MongoCollection<Document> collection = getPollsCollection();

            String pollJson = objectMapper.writeValueAsString(pollRequest);
            Document pollBsonDocument = Document.parse(pollJson);

            return RatpackMongoClient.insertOne(collection, pollBsonDocument)
                    .map(validation -> validation
                            .map((ignored) -> pollBsonDocument)
                            .map(this::mapBsonDocumentToDomainObject)
                    );

        } catch (JsonProcessingException e) {
            return error(e);
        }
    }

    public Promise<Validation<FindOneError, Poll>> retrievePoll(String pollId) {
        return RatpackMongoClient.findOneById(getPollsCollection(), pollId)
                .map(validation -> validation.map(this::mapBsonDocumentToDomainObject));
    }

    private Poll mapBsonDocumentToDomainObject(Document document) {
        //noinspection unchecked
        return new Poll(
                document.getObjectId("_id").toHexString(),
                document.getString("topic"),
                (List) document.get("options")
        );
    }

    private MongoCollection<Document> getPollsCollection() {
        MongoClient mongo = MongoClients.create("mongodb://localhost:12345");
        MongoDatabase database = mongo.getDatabase("webpoll");
        return database.getCollection("polls");
    }
}

File: src/main/java/myratpackexamples/webpoll/PollRepository.kt
package myratpackexamples.webpoll

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.inject.Inject
import com.mongodb.async.client.MongoClients
import com.mongodb.async.client.MongoCollection
import io.vavr.control.Validation
import io.vavr.control.Validation.invalid
import myratpackexamples.webpoll.InsertOneError.InsertOneJsonProcessingError
import myratpackexamples.webpoll.PollResponse.*
import org.bson.Document
import ratpack.exec.Promise
import ratpack.exec.Promise.value

class PollRepository @Inject constructor(val objectMapper: ObjectMapper) {

    private val pollsCollection: MongoCollection<Document>
        get() {
            val mongo = MongoClients.create("mongodb://localhost:12345")
            val database = mongo.getDatabase("webpoll")
            return database.getCollection("polls")
        }

    fun storePoll(poll: PollEntity.Poll): Promise<Validation<InsertOneError, Poll>> {
        try {
            val pollJson = objectMapper.writeValueAsString(poll)
            val pollBsonDocument = Document.parse(pollJson)

            return pollsCollection.insertOne(pollBsonDocument)
                    .map { it.map { _ -> pollBsonDocument }.map({ document -> mapBsonDocumentToDomainObject(document, null) }) }

        } catch (e: JsonProcessingException) {
            return value(invalid<InsertOneError, Poll>(InsertOneJsonProcessingError(e)))
        }

    }

    fun replacePoll(pollId: String?, poll: PollEntity.Poll): Promise<Validation<ReplaceOneError, Poll>> {
        try {
            val pollJson = objectMapper.writeValueAsString(poll)
            val pollBsonDocument = Document.parse(pollJson)

            return pollsCollection.replaceOne(pollId, pollBsonDocument)
                    .map { it.map { _ -> pollBsonDocument }.map { document -> mapBsonDocumentToDomainObject(document, pollId) } }

        } catch (e: JsonProcessingException) {
            return value(invalid<ReplaceOneError, Poll>(ReplaceOneError.ReplaceOneJsonProcessingError(e)))
        }

    }

    fun retrievePoll(pollId: String?): Promise<Validation<FindOneError, Poll>> {
        return pollsCollection.findOneById(pollId)
                .map { it.map { document -> mapBsonDocumentToDomainObject(document, null) } }
    }

    private fun mapBsonDocumentToDomainObject(document: Document, pollId: String?): Poll {
        @Suppress("UNCHECKED_CAST")
        return Poll(
                id = pollId ?: document.getObjectId("_id").toHexString(),
                topic = document.getString("topic"),
                options = document["options"] as MutableList<String>,
                votes = mapVotes(document)
        )
    }

    private fun mapVotes(document: Document): List<Vote> {
        val votes = document["votes"]
        if (votes == null) return emptyList()

        @Suppress("UNCHECKED_CAST")
        return (votes as MutableList<Document>).map {
            Vote(
                    voter = it.getString("voter"),
                    selections = mapSelections(it)
            )
        }
    }

    private fun mapSelections(document: Document): List<Selection> {
        val selections = document["selections"]
        if (selections == null) return emptyList()

        @Suppress("UNCHECKED_CAST")
        return (selections as MutableList<Document>).map {
            Selection(
                    option = it.getString("option"),
                    selected = Selected.valueOf(it.getString("selected"))
            )
        }
    }
}
